rai/node/wallet.cpp

```diff
@@ -925,43 +929,70 @@ std::shared_ptr<rai::block> rai::wallet::change_action (rai::account const & sou
 }
 
-std::shared_ptr<rai::block> rai::wallet::send_action (rai::account const & source_a, rai::account const & account_a, rai::uint128_t const & amount_a, bool generate_work_a)
+std::shared_ptr<rai::block> rai::wallet::send_action (rai::account const & source_a, rai::account const & account_a, rai::uint128_t const & amount_a, bool generate_work_a, boost::optional<std::string> id_a)
 {
 	std::shared_ptr<rai::block> block;
+	bool cached_block = false;
 	{
-		rai::transaction transaction (store.environment, nullptr, false);
-		if (store.valid_password (transaction))
+		rai::transaction transaction (store.environment, nullptr, (bool) id_a);
+		if (auto id = id_a)
 		{
-			auto existing (store.find (transaction, source_a));
-			if (existing != store.end ())
+			for (auto i = rai::store_iterator (transaction, store.send_action_ids), n = rai::store_iterator (nullptr); i != n; ++i)
 			{
-				auto balance (node.ledger.account_balance (transaction, source_a));
-				if (!balance.is_zero () && balance >= amount_a)
+				if (i->first.size () == id->size () && memcmp (i->first.data (), id->data (), id->size ()) == 0)
 				{
-					rai::account_info info;
-					auto error1 (node.ledger.store.account_get (transaction, source_a, info));
-					assert (!error1);
-					rai::raw_key prv;
-					auto error2 (store.fetch (transaction, source_a, prv));
-					assert (!error2);
-					block.reset (new rai::send_block (info.head, account_a, balance - amount_a, prv, source_a, generate_work_a ? work_fetch (transaction, source_a, info.head) : 0));
+					rai::bufferstream stream (reinterpret_cast<uint8_t const *> (i->second.data ()), i->second.size ());
+					block = rai::deserialize_block (stream, rai::block_type::send);
+					cached_block = true;
+					node.network.republish_block (transaction, block);
+					break;
+				}
+			}
+		}
+		if (block == nullptr)
+		{
+			if (store.valid_password (transaction))
+			{
+				auto existing (store.find (transaction, source_a));
+				if (existing != store.end ())
+				{
+					auto balance (node.ledger.account_balance (transaction, source_a));
+					if (!balance.is_zero () && balance >= amount_a)
+					{
+						rai::account_info info;
+						auto error1 (node.ledger.store.account_get (transaction, source_a, info));
+						assert (!error1);
+						rai::raw_key prv;
+						auto error2 (store.fetch (transaction, source_a, prv));
+						assert (!error2);
+						block.reset (new rai::send_block (info.head, account_a, balance - amount_a, prv, source_a, generate_work_a ? work_fetch (transaction, source_a, info.head) : 0));
+					}
+				}
+			}
+			if (block != nullptr)
+			{
+				if (auto id = id_a)
+				{
+					std::vector<uint8_t> vector;
+					{
+						rai::vectorstream stream (vector);
+						block->serialize (stream);
+					}
+					auto status (mdb_put (transaction, store.send_action_ids, rai::mdb_val (id->size (), const_cast<char*> (id->data ())), rai::mdb_val (vector.size (), vector.data ()), 0));
+					assert (status == 0);
```

Comment:
Shouldn't `send_action` propagate errors to the caller?  In release mode, mdb_put errors are swallowed.

Reply:
Fixed